python/dolfinx/io/gmshio.py

```diff
@@ -330,7 +373,43 @@ def model_to_mesh(
     else:
         ft = meshtags(mesh, tdim - 1, np.empty(0, dtype=np.int32), np.empty(0, dtype=np.int32))
 
-    return (mesh, ct, ft)
+    if has_edge_data:
+        # Permute edges from MSH to DOLFINx ordering
+        edge_type = _cpp.mesh.cell_entity_type(
+            _cpp.mesh.to_type(str(ufl_domain.ufl_cell())), tdim - 2, 0
+        )
+        gmsh_edge_perm = cell_perm_array(edge_type, num_edge_nodes)
+        marked_edges = marked_edges[:, gmsh_edge_perm]
+
+        local_entities, local_values = distribute_entity_data(
+            mesh, tdim - 2, marked_edges, edge_values
+        )
+        mesh.topology.create_connectivity(topology.dim - 2, tdim)
+        adj = adjacencylist(local_entities)
+        et = meshtags_from_entities(mesh, tdim - 2, adj, local_values.astype(np.int32, copy=False))
+        et.name = "Edge tags"
+    else:
+        et = meshtags(mesh, tdim - 2, np.empty(0, dtype=np.int32), np.empty(0, dtype=np.int32))
```

Comment:
Is there a good reason not to use None?

Reply:
This is a choice to be made. I think using `None` make sense and would signal that there is no physical group for a given dimension. However, it would make the return type `MeshTags | None` which would complicate the type checking further down (i.e the user need to check for `None` before using the attributes). I think both ways are OK, so I leave it to you to decide.

Reply:
We're trying to remove sentinel types from the interface, i.e. return values that imply None. I'd prefer to return None. @jorgensd ?

Reply:
I wonder if there ever is a case where one want to distinguish between having any data on a local process and there not being tags at all in the Gmsh model.

I would prefer a monotype return value, ie meshtags (if the above comment doesn’t raise any other concerns)

Reply:
OK, I guess I'm unsure now. If we're sticking with this, is this behaviour documented?

Reply:
A starting point is that in parallel all processes must return the same type. 

If the Gmsh model has no tags, then the return type should be `None` because there are no tags. If the user expects tags and gets `None`, then they know something has gone wrong. Getting empty `MeshTags` because there are no tags but tags were expected is a problem.

In parallel, if I get empty tags on one process I don't know if that means that the mesh has no tags or if the portion of the mesh on the current process has no tags but that other portions might.

Reply:
One problem with using `None` is that you would need to add checks everywhere if you want to make the type checker happy and this can make the code rather verbose. For example the following code
```python
mesh_data = gmshio.model_to_mesh(model, comm, rank=0)
mesh_data.mesh.name = name
mesh_data.cell_tags.name = f"{name}_cells"
mesh_data.facet_tags.name = f"{name}_facets"
mesh_data.edge_tags.name = f"{name}_edges"
mesh_data.vertex_tags.name = f"{name}_vertices"
```
will need to be rewritten as
```python
mesh_data = gmshio.model_to_mesh(model, comm, rank=0)
mesh_data.mesh.name = name
assert mesh_data.cell_tags is not None
mesh_data.cell_tags.name = f"{name}_cells"
assert mesh_data.facet_tags is not None
mesh_data.facet_tags.name = f"{name}_facets"
assert mesh_data.edge_tags is not None
mesh_data.edge_tags.name = f"{name}_edges"
assert mesh_data.vertex_tags is not None
mesh_data.vertex_tags.name = f"{name}_vertices"
```
to make `mypy` happy (or some other type of check). Not saying one is better than this other, but just to be aware that users downstream would need add these checks if they type check their code (which also includes the demos). 

Another option could be to add an argument `default_tag` which will be used if no tags are found, e.g `mesh_data = gmshio.model_to_mesh(model, comm, rank=0, default_tag=0)`

Reply:
I think the checker is a secondary consideration. Don't want the tail wagging the dog :).

The asserts are not necessarily bad; if a user is expected mesh tags from the mesh, then the assertions check that they got them.

The most important point is that empty `MeshTags` and `None` are not the same thing, and points like this can be critical in parallel. If a rank has an empty `MeshTags` then other ranks may non-empty have tags. If a rank has `None`, then the mesh has no mesh tags, i.e. no ranks will have tags. Knowing this can be important for handling collective operations.

Reply:
Tags are now set to `None` if they don't exist in https://github.com/FEniCS/dolfinx/pull/3547/commits/58ae9057d4359344b1844c20f3074677ba0f4f09.